Move client/unversioned/remotecommand to client-go

dshulyak opened the pull request:
Module remotecommand originally part of kubernetes/pkg/client/unversioned was moved
to client-go/tools, and will be used as authoritative in kubectl, e2e and other places.

Module remotecommand relies on util/exec module which was copied to client-go/pkg/util

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41331)
<!-- Reviewable:end -->

dshulyak commented:
@caesarxuchao Can you take a look please?

dshulyak commented:
@caesarxuchao Hello, can you tell me when you will have time to take a look at this PR? I have some weird results, for example apimachinery is removed from _vendor, and i am not able to figure out whether i should add it back manually or not..

caesarxuchao commented on staging/copy.sh:
```diff
@@ -125,6 +127,7 @@ rm -rf "${CLIENT_REPO_TEMP}"/vendor/k8s.io/kubernetes
 # client-go will share the vendor of the main repo for now. When client-go
 # becomes a standalone repo, it will have its own vendor
 mv "${CLIENT_REPO_TEMP}"/vendor "${CLIENT_REPO_TEMP}"/_vendor
+rm -r "${CLIENT_REPO_TEMP}"/_vendor/github.com/docker/docker/project/
```
Why do you need to manually remove this folder?

dshulyak replied:
this folder containers symlink that references file that is not copied

caesarxuchao commented on staging/src/k8s.io/client-go/cmd/kubeadm/app/apis/kubeadm/doc.go:
```diff
@@ -0,0 +1,18 @@
+/*
```
Why does kubeadm get imported?

caesarxuchao commented on staging/src/k8s.io/client-go/federation/apis/federation/doc.go:
```diff
@@ -0,0 +1,17 @@
+/*
```
Looks like remotecommand imports federation types as well, can we remove the dependency on federation and kubeadm types?

dshulyak replied:
yes, looks like i can

caesarxuchao commented on staging/src/k8s.io/client-go/pkg/kubelet/qos/doc.go:
```diff
@@ -0,0 +1,25 @@
+/*
```
We are trying to remove the packages in /pkg/kubelet,  which are are duplicate packages between kuberentes and client-go. Are there other users of `pkg/kubelet/qos` except for client-go? If so, we can move this package to client-go, and remove the original copy. cc @deads2k @sttts

deads2k commented:
No more copies.  If this needs to move, we aren't going to make a copy of it.  The authoritative (and only) copy should be in staging/client-go.

caesarxuchao commented:
@dshulyak i think it's correct that apimachinery is removed from vendor/. deads2k ran copy.sh once yesterday. You can rebase your PR so that you'll see less changes.

sttts commented:
Why is remotecommand needed in client-go? Everything we move there we effectively declare to be a public, stable API (at least within minor versions, but still). Hence, what is the use-case?

Moreover, `kubelet/server/remotecommand` and `pkg/util/term` with those tons of dependencies move too?

IMO we need better separation, i.e. a number of dependencies cut before moving it.

dshulyak commented:
@sttts I am using exec in several k8s related projects to run different tests. I think there is definitely demand for such thing in client https://github.com/kubernetes/client-go/issues/45

I can check what can be done to cut those dependencies before moving it

sttts commented:
@dshulyak please try to cut those terminal related dependencies and of course the kubelet dep.

sttts commented on pkg/util/remotecommand/constants.go:
```diff
@@ -0,0 +1,53 @@
+/*
```
pkg/util is an odd location for this. It's actually part of the API. As we don't have a better location for this, moving it to client-go would be fine IMO.

dshulyak replied:
the same for pkg/util/exec, i can't find good place for it

dshulyak replied:
perhaps pkg/util/remotecommand should be in apimachinery, i see similar comment for port forwarding from @deads2k

dshulyak commented:
@sttts @deads2k @caesarxuchao I think this change now in good shape. Please review it when you will have time.

I think I need to move pkg/util/remotecommand/constants.go to apimachinery (the same as portforward/constants.go) and unify it with kubelet. I will do it in another change. Also I am not sure what is the proper place for pkg/util/exec, so I just placed it in client-go repo

dshulyak commented:
@k8s-bot test this

dshulyak commented:
I moved part of this PR into separate one, for easier review, https://github.com/kubernetes/kubernetes/pull/41543. Will rebase afterwards

caesarxuchao commented:
Sorry I have been ignoring the PR. I'm taking another look now.

caesarxuchao commented:
I left a few comments in https://github.com/kubernetes/kubernetes/pull/41543. Let's get that merged first.

dshulyak commented:
@k8s-bot bazel test this

dshulyak committed:
Move client/unversioned/remotecommand to client-go

Module remotecommand originally part of kubernetes/pkg/client/unversioned was moved
to client-go/tools, and will be used as authoritative in kubectl, e2e and other places.

Module remotecommand relies on util/exec module which will be copied to client-go/pkg/util

dshulyak commented:
@caesarxuchao @lavalamp Squashed followup for https://github.com/kubernetes/kubernetes/pull/41543

caesarxuchao commented on staging/src/k8s.io/client-go/pkg/util/exec/exec.go:
```diff
@@ -0,0 +1,188 @@
+limitations under the License.
+*/
+
+package exec
```
Can we remove the original file? Also, if we can, please move it to `client-go/util`, not `client-go/pkg/util`, otherwise copy.sh doesn't work proplerly (see this [line](https://github.com/kubernetes/kubernetes/blob/master/staging/copy.sh#L100)).

dshulyak replied:
I noticed that module `pkg/util/parsers` was used in client-go in the same way, thats way i copied `pkg/util/exec` instead of moving. Also i wasnt sure that `pkg/util/exec` belongs to client-go

dshulyak replied:
But if you are sure that it is a way to go, i can do it ofcourse

dshulyak replied:
instead of removing pkg/util/exec from kubernetes repo, i extracted the only used object exec.CodeExitError to util/exec in client-go repo

caesarxuchao commented:
Also please run hack/update-staging-client-go.sh.  Run it after you move the `exec` package, otherwise that code will be wiped out.

Otherwise lgtm.

caesarxuchao commented:
@dshulyak are you still working on this? Or do you need any help?

dshulyak commented:
@caesarxuchao Sorry, i got a bit distracted with other stuff. I will address mentioned issues in a day or two

dshulyak committed:
Add exec module to client-go/util only with required code

client-go/pkt/util was removed in favor of client-go util, which consists
only from CodeExitError and ExitError interface

dshulyak committed:
Fix golint and godeps verification

dshulyak committed:
Add remotecommand to client-go/pkg/api visibility rules

caesarxuchao commented on staging/src/k8s.io/client-go/pkg/api/BUILD:
```diff
@@ -26,6 +26,7 @@ go_library(
     visibility = [
         "//vendor/k8s.io/client-go/pkg/api:__subpackages__",
         "//vendor/k8s.io/client-go/pkg/apis:__subpackages__",
+        "//vendor/k8s.io/client-go/tools/remotecommand:__subpackages__",
```
Ah, i don't want to add more dependency to client-go/pkg/api, as we are trying to get rid of this package. Let me check why remotecommand depends on pkg/api.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/remotecommand/v1.go:
```diff
@@ -24,7 +24,7 @@ import (
 	"github.com/golang/glog"
 	"k8s.io/apimachinery/pkg/util/httpstream"
-	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/client-go/pkg/api"
```
It looks like it the api package is needed only for some constants. These constants should exist in the api/v1 package as well, could you use the copies there? Let's copy those constants to api/v1 if they are not already there

caesarxuchao commented:
@dshulyak thanks! Could you get rid of the dependency on client-go/pkg/api? Otherwise lgtm. Sorry the  dependencies are a mess.

dshulyak committed:
Use streaming constants from pkg/api/v1 instead of pkg/api

caesarxuchao commented:
/release-note-none

caesarxuchao commented:
/lgtm
/approve

caesarxuchao commented:
Thanks @dshulyak !

caesarxuchao commented on staging/src/k8s.io/client-go/util/exec/exec.go:
```diff
@@ -0,0 +1,52 @@
+// ExitError is an interface that presents an API similar to os.ProcessState, which is
+// what ExitError from os/exec is.  This is designed to make testing a bit easier and
+// probably loses some of the cross-platform properties of the underlying library.
+type ExitError interface {
```
@dshulyak let me know if you are up for a follow-up.

dshulyak replied:
@caesarxuchao i copied only parts that were used in remotecommand, do you think it makes sense to move the whole `pkg/util/exec`? if so, i can do it ofcourse

caesarxuchao replied:
I see. Can we move `ExitError` to the remotecmd package? (removing the copy in client-go/util/exec/exec.go and in kubernetes/util/exec/exec.go). In a followup PR is fine. There are not many references to ExitError:
```bash
$ git grep -l "CodeExitError" | grep -v pkg/util/exec
pkg/client/unversioned/remotecommand/v4.go
pkg/kubectl/cmd/run.go
pkg/kubectl/cmd/util/helpers_test.go
pkg/kubelet/remote/remote_runtime.go
test/e2e/framework/util.go
```

caesarxuchao replied:
@dshulyak please let me know if you are interested to do the follow up

dshulyak replied:
@caesarxuchao sure, i will do it

caesarxuchao replied:
Thanks, please assign me to the PR when it's out.

lavalamp commented:
/approve

caesarxuchao commented:
@k8s-bot unit test this

caesarxuchao commented:
@k8s-bot gce etcd3 e2e test this

caesarxuchao commented:
@k8s-bot kops aws e2e test this

caesarxuchao commented:
/lgtm
